### PR TITLE
How to prevent deprecated type aliases in numpy?

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

(Details)

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/test_no_numpy_builtin_type_aliases.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    for direcory in ["mlflow", "tests"]:
+        for path in iterate_python_scripts(direcory):
+            finder.find(path)
+
+    msg = (
+        "NumPy's aliases for built-in types (e.g. np.int) have been deprecated. "
+        "Please replace them by following this instruction: "
+        "https://numpy.org/devdocs/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated\n"  # noqa
+    )
+    assert finder.found == [], msg + finder.summary()
```
Comment (harupy) [inline tests/test_no_numpy_builtin_type_aliases.py:88]: The result looks like:

https://github.com/mlflow/mlflow/pull/4098/checks?check_run_id=1885889164#step:4:1673


```
E       AssertionError: numpy's built-in type aliases (e.g. np.int) have been deprecated. Please replace them by following this instruction: https://numpy.org/devdocs/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated
E         mlflow/types/schema.py:16:15 | np.object
E         mlflow/types/schema.py:48:50 | np.object
E         mlflow/pyfunc/__init__.py:299:23 | np.object
E         mlflow/pyfunc/__init__.py:302:48 | np.object
E         mlflow/pyfunc/__init__.py:814:83 | np.int
E         mlflow/pyfunc/__init__.py:814:91 | np.long
E         mlflow/utils/proto_json_utils.py:64:26 | np.object
E         tests/types/test_schema.py:144:59 | np.object
E         tests/types/test_schema.py:146:66 | np.object
E         tests/types/test_schema.py:148:76 | np.object
E         tests/types/test_schema.py:150:66 | np.object
E         tests/types/test_schema.py:152:65 | np.object
E         tests/types/test_schema.py:160:59 | np.str
E         tests/types/test_schema.py:164:60 | np.bool
E         tests/tensorflow/test_tensorflow2_model_export.py:250:43 | np.float
E         tests/tensorflow/test_tensorflow2_model_export.py:251:65 | np.float
E         tests/pyfunc/test_scoring_server.py:437:61 | np.object
E         tests/pyfunc/test_scoring_server.py:440:38 | np.bool
E         tests/pyfunc/test_spark.py:22:19 | np.int
E         tests/pyfunc/test_spark.py:22:27 | np.str
E         tests/pyfunc/test_spark.py:96:29 | np.int
E         tests/pyfunc/test_model_export_with_loader_module_and_data_path.py:150:14 | np.object
E         tests/pyfunc/test_model_export_with_loader_module_and_data_path.py:263:31 | np.object
E         tests/pyfunc/test_model_export_with_loader_module_and_data_path.py:264:31 | np.object
E         tests/pyfunc/test_model_export_with_loader_module_and_data_path.py:265:31 | np.object
E         tests/pyfunc/test_model_export_with_loader_module_and_data_path.py:266:31 | np.object
E         tests/pyfunc/test_model_export_with_loader_module_and_data_path.py:267:31 | np.object
E         tests/shap/test_shap.py:74:36 | np.float
E         tests/statsmodels/model_fixtures.py:156:80 | np.int
E         tests/statsmodels/model_fixtures.py:157:82 | np.int
E         tests/statsmodels/model_fixtures.py:158:82 | np.int
E         tests/onnx/test_onnx_model_export.py:168:35 | np.bool
E         tests/onnx/test_onnx_model_export.py:170:68 | np.bool
```

**Comment by aloosley** [comment]
Hi @harupy , it seems like a much more general solution that works for other warnings is to filter warnings that occur during testing and raise a testing error if any of a certain type with a specific message occur.  

For example, the first entry to such a warning-filter would be any `DeprecationWarning` where `"is a deprecated alias for the builtin" in msg and "np" in msg`.

**Comment by harupy** [comment]
Sounds like a nice approach. Does that approach works when third-party libraries (e.g. pandas?) contain the deprecated aliases?

**Comment by harupy** [comment]
```
# pytest.ini

filterwarnings =
    error:.*is a deprecated alias for the builtin.*:DeprecationWarning:mlflow
```

may do the trick.

**Comment by aloosley** [comment]
If you want to add this one DeprecationWarning test, you might as well make tests fail on any DeprecationWarning because anybody can right any call to any deprecated code in any PR and old calls are deprecated all the time.  However, this design choice obviates the point of having DeprecationWarnings.  That is, your code will break in the future but for now everything is okay, and you have time to make changes.  On the other hand, writing required tests that catch DeprecationWarnings robs package devs of the time gifted by the Deprecators to make changes.  Furthermore, having such a warnings catchall will cause all PRs to be blocked until all new warnings created by devs working on ML and all dependencies are handled.  

The focus should therefore be to capture and track warnings, for example, a non-required catch-all (or all with exceptions) warnings test.  Following such a design thought, I didn't write any test for the deprecated numpy calls PR.  You know what I mean?

**Comment by harupy** [comment]
> you might as well make tests fail on any DeprecationWarning

not any DeprecationWarning, right?

**Comment by aloosley** [comment]
> @aloosley Thanks for the comment!
> 
> > you might as well make tests fail on any DeprecationWarning
> 
> Not any `DeprecationWarning`, right? `error:.*is a deprecated alias for the builtin.*:DeprecationWarning:mlflow` means turning warnings that satisfy the following conditions into errors when running `pytest`.
> 
>     1. is `DeprecationWarning`
> 
>     2. warning message matches `.*is a deprecated alias for the builtin.*` (we can use a more strict regular expression)
> 
>     3. originates from `mlflow` (please see the example below)
> 
> 
> ```python
> # mlflow/xxx/yyy.py
> 
> a = np.object 
> # this is considered as an error
> 
> b = some_function_imported_from_pandas_that_contains_np_object()
> # this isn't because the warning is emitted from pandas
> ```
> 
> ```python
> # pandas/xxx/yyy.py
> 
> def some_function_imported_from_pandas_that_contains_np_object(...):
>     a = np.object
> ```
I like your criteria, looks good :+1: !

**Comment by harupy** [comment]
#5286 fixed this issue

**Pull request closed without merging**